hudi-client/hudi-java-client/src/main/java/org/apache/hudi/table/action/commit/JavaDeleteHelper.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.hudi.exception.HoodieUpsertException;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.hudi.table.WorkloadProfile;
+import org.apache.hudi.table.WorkloadStat;
+import org.apache.hudi.table.action.HoodieWriteMetadata;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@SuppressWarnings("checkstyle:LineLength")
+public class JavaDeleteHelper<R> extends
+    AbstractDeleteHelper<EmptyHoodieRecordPayload, List<HoodieRecord<EmptyHoodieRecordPayload>>, List<HoodieKey>, List<WriteStatus>, R> {
+
+  private JavaDeleteHelper() {
+  }
+
+  private static class DeleteHelperHolder {
+    private static final JavaDeleteHelper JAVA_DELETE_HELPER = new JavaDeleteHelper();
+  }
+
+  public static JavaDeleteHelper newInstance() {
+    return DeleteHelperHolder.JAVA_DELETE_HELPER;
+  }
+
+  @Override
+  public List<HoodieKey> deduplicateKeys(List<HoodieKey> keys, HoodieTable<EmptyHoodieRecordPayload, List<HoodieRecord<EmptyHoodieRecordPayload>>, List<HoodieKey>, List<WriteStatus>> table, int parallelism) {
+    boolean isIndexingGlobal = table.getIndex().isGlobal();
+    if (isIndexingGlobal) {
+      HashSet<String> recordKeys = keys.stream().map(HoodieKey::getRecordKey).collect(Collectors.toCollection(HashSet::new));
+      List<HoodieKey> deduplicatedKeys = new LinkedList<>();
+      keys.forEach(x -> {
+        if (recordKeys.contains(x.getRecordKey())) {
+          deduplicatedKeys.add(x);
+        }
+      });
+      return deduplicatedKeys;
+    } else {
+      HashSet<HoodieKey> set = new HashSet<>(keys);
+      keys.clear();
+      keys.addAll(set);
+      return keys;
+    }
+  }
+
+  @Override
+  public HoodieWriteMetadata<List<WriteStatus>> execute(String instantTime,
+                                                        List<HoodieKey> keys,
+                                                        HoodieEngineContext context,
+                                                        HoodieWriteConfig config,
+                                                        HoodieTable<EmptyHoodieRecordPayload, List<HoodieRecord<EmptyHoodieRecordPayload>>, List<HoodieKey>, List<WriteStatus>> table,
+                                                        BaseCommitActionExecutor<EmptyHoodieRecordPayload, List<HoodieRecord<EmptyHoodieRecordPayload>>, List<HoodieKey>, List<WriteStatus>, R> deleteExecutor) {
+    try {
+      HoodieWriteMetadata<List<WriteStatus>> result = null;
+      List<HoodieKey> dedupedKeys = keys;
+      final int parallelism = config.getDeleteShuffleParallelism();
+      if (config.shouldCombineBeforeDelete()) {
+        // De-dupe/merge if needed
+        dedupedKeys = deduplicateKeys(keys, table, parallelism);
+      }
+
+      List<HoodieRecord<EmptyHoodieRecordPayload>> dedupedRecords =
```

Comment:
will it also work with `_hoodie_is_deleted` field set to `true` in HoodieRecord?

Reply:
It seems "_hoodie_is_deleted" is supported in upsert, the delete operator only need List<HoodieKey> keys.